src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -146,199 +162,417 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
 	{
-		EnsureTableNotDistributed(relationId);
+		return CreateReferenceTableColocationId();
 	}
 	else
 	{
-		EnsureTableNotDistributed(relationId);
-		EnsureLocalTableEmpty(relationId);
-	}
+		/*
+		 * Get an exclusive lock on the colocation system catalog. Therefore, we
+		 * can be sure that there will no modifications on the colocation table
+		 * until this transaction is committed.
+		 */
+		Relation pgDistColocation = heap_open(DistColocationRelationId(), ExclusiveLock);
 
-	colocationId = CreateReferenceTableColocationId();
+		Oid distributionColumnType = distributionColumn->vartype;
 
-	/* first, convert the relation into distributed relation */
-	ConvertToDistributedTable(relationId, distributionColumnName,
-							  DISTRIBUTE_BY_NONE, REPLICATION_MODEL_2PC, colocationId);
+		if (pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) == 0)
+		{
+			/* check for default colocation group */
+			colocationId = ColocationId(ShardCount, ShardReplicationFactor,
+										distributionColumnType);
+
+			if (colocationId == INVALID_COLOCATION_ID)
+			{
+				colocationId = CreateColocationGroup(ShardCount, ShardReplicationFactor,
+													 distributionColumnType);
+			}
+		}
+		else if (pg_strncasecmp(colocateWithTableName, "none", NAMEDATALEN) == 0)
+		{
+			colocationId = GetNextColocationId();
+		}
+		else
+		{
+			text *colocateWithTableNameText = cstring_to_text(colocateWithTableName);
+			Oid sourceRelationId = ResolveRelationId(colocateWithTableNameText);
 
-	/* now, create the single shard replicated to all nodes */
-	CreateReferenceTableShard(relationId);
+			EnsureTableCanBeColocatedWith(relationId, replicationModel,
+										  distributionColumnType, sourceRelationId);
 
-	CreateTableMetadataOnWorkers(relationId);
+			colocationId = TableColocationId(sourceRelationId);
+		}
 
-	/* copy over data for regular relations */
-	if (relationKind == RELKIND_RELATION)
-	{
-		CopyLocalDataIntoShards(relationId);
+		heap_close(pgDistColocation, NoLock);
 	}
+
+	return colocationId;
 }
 
 
 /*
- * ConvertToDistributedTable converts the given regular PostgreSQL table into a
- * distributed table. First, it checks if the given table can be distributed,
- * then it creates related tuple in pg_dist_partition.
+ * EnsureRelationCanBeDistributed checks whether Citus can safely distribute given table
+ * with the given configuration. We perform almost all safety checks for distributing
+ * table here. If there is an unsatisfied requirement, we error out and do not
+ * distribute the table.
  *
- * XXX: We should perform more checks here to see if this table is fit for
- * partitioning. At a minimum, we should validate the following: (i) this node
- * runs as the master node, (ii) table does not make use of the inheritance
- * mechanism and (iii) table does not have collated columns.
+ * This function assumes, callers already took necessary locks to ensure there will not
+ * be any change in the given table.
```

Comment:
` given table.` -> `given relation` ?
`callers already ` - >`the callers have already acquired` ?